test/cql-pytest/test_compaction_tombstone_gc.py

```diff
@@ -0,0 +1,141 @@
+# Copyright 2024-present ScyllaDB
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+from util import new_materialized_view, new_test_table
+import nodetool
+import time
+
+def test_tombstone_gc_with_conflict_in_memtable(scylla_only, cql, test_keyspace):
+    """
+    Regression test for fixed https://github.com/scylladb/scylladb/issues/20423
+    """
+    schema = "k int, v int, primary key (k, v)"
+    with new_test_table(cql, test_keyspace, schema, extra="with gc_grace_seconds = 0") as table:
+        with nodetool.no_autocompaction_context(cql, table):
+            nodetool.setlogginglevel(cql, "compaction", "trace")
+            # Insert initial data into the base table:
+            # Row 1 is expected to be garbage collected after expiration
+            cql.execute(f"insert into {table} (k, v) values (1, 1) using timestamp 1 and ttl 1")
+            cql.execute(f"insert into {table} (k, v) values (1, 2) using timestamp 2")
+            nodetool.flush(cql, table)
+            # Delete row 2, this deletion is expected to be kept
+            # when we insert backdated live data into the memtable
+            cql.execute(f"delete from {table} using timestamp 3 where k=1 and v=2")
+            # Flush all tables explicitly now, since this is skipped in the test on purpose by next major compaction.
+            nodetool.flush_all(cql)
+
+            # let gc_grace_seconds and ttl pass
+            time.sleep(2)
```

Comment:
This is going to be flaky. You can't rely on the time running at a certain speed in the test environment. Suggest to replace with a loop that sleeps and checks the condition the test expects.

Reply:
Actually, if you create a cell with ttl=1 and then sleep(2), you are guaranteed to have this item expired. CQL's ttl does not require anything to run (unlike Alternator's ttl) - the expiration is just checked any time this value is read, and you'll always see the item has expired.

I think that sleep(1) would have been guaranteed enough as well.

However I claimed above that this test probably doesn't really need a TTL for this test. But maybe Benny wanted a TTL to test that case as well?

Reply:
hmm, both `ttl` and `gc_grace_seconds` refer to system time.
why isn't `time.sleep` reliable?

Reply:
time.sleep() can sleep an arbitrary amount of wall clock time, by definition of the call.

Reply:
@nyh if we're not going to see the rows thanks to CQL what is the test testing?

Reply:
@kostja, sleep() will sleep **at least** two seconds. Unless the machine is crazy over-loaded, it will be a little bit over 2 seconds. If a 2-second sleep took 20 seconds, we have a serious infrastructure problem. So in general, a 2 second sleep is not an "unbounded" sleep. But it's still a **long** sleep - 100 2-second sleeps in 100 places take together 200 seconds. In test/alternator, each sleep() caused me personal pain (it's not easy to maintain the situation that running almost 1000 tests sequentially takes a total of 30 seconds). 

In cql-pytest I'm less uptight about it, but I still want to carefully consider each sleep and why it's needed:
1. In this case the 2-second sleep is needed because this test has TTL, and I wondered, why it needs TTL at all to demonstrate its point. But I agreed that if Benny thinks that TTL makes this test even harder somehow, and better, then I'm ok that he uses it.
2. I didn't understand why gc_grace_seconds=0 requires 1 second of sleep and not 0. I still don't understand if it's a bug or intentional.
3. I didn't understand what is that extra commit_gc thing and how it also requires a 1 second / 2 second sleep.

Reply:
> @nyh if we're not going to see the rows thanks to CQL what is the test testing?

Originally, @bhalevy tested the sstables on disk. But now he uses @denesb 's "SELECT MUTATION FRAGMENTS" feature which is much simpler (in my opinion). All of this is Scylla-specific (these tests cannot be run on Cassandra).